src/TestFramework/TestFramework/Assertions/Assert.That.cs

```diff
@@ -0,0 +1,786 @@
+                        string argumentsList = string.Join(", ", arguments);
+#if NET
+                        result.Append(CultureInfo.InvariantCulture, $"new {collectionType}<{genericType}> {{ {argumentsList} }}");
+#else
+                        result.Append($"new {collectionType}<{genericType}> {{ {argumentsList} }}");
+#endif
+                        i = braceEnd;
+                        continue;
+                    }
+                }
+            }
+
+            result.Append(input[i]);
+            i++;
+        }
+
+        return result.ToString();
+    }
+
+    private static bool TryMatchListInitPattern(string input, int startIndex, out string collectionType, out int patternEnd)
+    {
+        collectionType = string.Empty;
+        patternEnd = startIndex;
+
+        // Check for "new " at the start
+        if (startIndex + 4 >= input.Length || !input.Substring(startIndex, 4).Equals("new ", StringComparison.Ordinal))
+        {
+            return false;
+        }
+
+        int pos = startIndex + 4;
+
+        // Skip whitespace
+        while (pos < input.Length && char.IsWhiteSpace(input[pos]))
+        {
+            pos++;
+        }
+
+        // Check for collection type names
+        string[] collectionTypes = ["List", "IList", "ICollection", "IEnumerable"];
+        string matchedType = string.Empty;
+
+        foreach (string type in collectionTypes)
+        {
+            if (pos + type.Length < input.Length &&
+                input.Substring(pos, type.Length).Equals(type, StringComparison.Ordinal))
+            {
+                matchedType = type;
+                pos += type.Length;
+                break;
+            }
+        }
+
+        if (string.IsNullOrEmpty(matchedType))
+        {
+            return false;
+        }
+
+        // Check for "`1()" pattern
+        if (pos + 4 >= input.Length || !input.Substring(pos, 4).Equals("`1()", StringComparison.Ordinal))
+        {
+            return false;
+        }
+
+        pos += 4;
+
+        // Skip whitespace
+        while (pos < input.Length && char.IsWhiteSpace(input[pos]))
+        {
+            pos++;
+        }
+
+        // Check for opening brace
+        if (pos >= input.Length || input[pos] != '{')
+        {
+            return false;
+        }
+
+        collectionType = matchedType;
+        patternEnd = pos;
+        return true;
+    }
+
+    private static string CleanTypeName(string typeName)
+        => typeName switch
+        {
+            "Int32" => "int",
+            "Int64" => "long",
+            "Int16" => "short",
+            "Byte" => "byte",
+            "SByte" => "sbyte",
+            "UInt32" => "uint",
+            "UInt64" => "ulong",
+            "UInt16" => "ushort",
+            "Single" => "float",
+            "Double" => "double",
+            "Decimal" => "decimal",
+            "Boolean" => "bool",
+            "String" => "string",
+            "Char" => "char",
+            "Object" => "object",
```

Comment:
It feels biased towards C# specifically but fine for me. I'm not sure though if we could be messing up with the unlikely case of users declaring types with the same name as those in BCL but in a different namespace?